src/jit/codegenlegacy.cpp

```diff
@@ -19462,14 +19462,30 @@ regMaskTP CodeGen::genCodeForCall(GenTreePtr call, bool valUsed)
                         regNumber indCallReg;
 
                         case IAT_VALUE:
+                        {
                             //------------------------------------------------------
                             // Non-virtual direct calls to known addressess
                             //
                             // The vast majority of calls end up here....  Wouldn't
                             // it be nice if they all did!
                             CLANG_FORMAT_COMMENT_ANCHOR;
 #ifdef _TARGET_ARM_
-                            if (!arm_Valid_Imm_For_BL((ssize_t)addr))
+                            // We may use direct call for some of recursive calls
+                            // as we can safely estimate the distance from the call site to the top of the method
+                            const int MAX_PROLOG_SIZE = MAX_USHORT_SIZE_T;
+                            const int codeOffset      = getEmitter()->emitCurCodeOffset + getEmitter()->emitCurIGsize;
+                            if (compiler->gtIsRecursiveCall(call->AsCall()) &&
+                                MAX_PROLOG_SIZE + codeOffset + 4 < 16777216) // if code offset fits in imm24
+                            {
+                                getEmitter()->emitIns_Call(emitter::EC_FUNC_TOKEN, methHnd,
+                                                           INDEBUG_LDISASM_COMMA(sigInfo) NULL, args, retSize,
+                                                           gcInfo.gcVarPtrSetCur, gcInfo.gcRegGCrefSetCur,
+                                                           gcInfo.gcRegByrefSetCur, ilOffset, REG_NA, REG_NA, 0,
+                                                           0,     /* ireg, xreg, xmul, disp */
+                                                           false, /* isJump */
```

Comment:
I updated the changes as you commented. Actually it was copied from https://github.com/dotnet/coreclr/blob/master/src/jit/codegenlegacy.cpp#L19495. I also tried to change this but I get formatting error like below so I had to leave it. FYI.

```
clang-format: there are formatting errors in /mnt/j/workspace/dotnet_coreclr/master/x64_ubuntu_formatting_prtest/src/jit/codegenlegacy.cpp
19:42:39 At Line 19512 Before:
19:42:39                                                            gcInfo.gcRegByrefSetCur, ilOffset,
19:42:39                                                            REG_NA, REG_NA, 0, 0, // ireg, xreg, xmul, disp
19:42:39 After:
19:42:39                                                            gcInfo.gcRegByrefSetCur, ilOffset, REG_NA, REG_NA, 0, 0, // ireg, xreg, xmul, disp
19:42:39 At Line 19512 Before:
19:42:39                                                            gcInfo.gcRegByrefSetCur, ilOffset, REG_NA, REG_NA, 0, 0, // ireg, xreg, xmul, disp
19:42:39 After:
19:42:39                                                            gcInfo.gcRegByrefSetCur, ilOffset, REG_NA, REG_NA, 0,
19:42:39                                                            0, // ireg, xreg, xmul, disp
19:42:39 At Line 19513 Before:
19:42:39                                                            0, // ireg, xreg, xmul, disp
19:42:39 After:
19:42:39                                                            0,     // ireg, xreg, xmul, disp
19:42:39 At Line 19514 Before:
19:42:39                                                            false,                // isJump
19:42:39 After:
19:42:39                                                            false, // isJump
```